src/types/datetime.rs

```diff
@@ -156,11 +165,18 @@ impl PyDateTime {
         let args = PyTuple::new(py, &[timestamp, time_zone_info]);
 
         unsafe {
-            let ptr = (PyDateTimeAPI.DateTime_FromTimestamp)(
-                PyDateTimeAPI.DateTimeType,
-                args.as_ptr(),
-                ptr::null_mut(),
-            );
+            #[cfg(PyPy)]
+            let ptr = PyDateTime_FromTimestamp(args.as_ptr());
```

Comment:
It occurs to me that this may lead to a panic or segfault.

I believe the current PyPy implementation of `PyDateTime_FromTimestamp` *requires* that you call `PyDatetime_Import()` before calling this function. All the other functions don't have this requirement because it's done once at global scope in the implementation of `PyDatetimeAPI`, so dereferencing that object at all guarantees that Import() has been called.

I think maybe the easiest short-term solution for this is to create a wrapper function around `PyDateTime_FromTimestamp` in the `ffi::datetime` module that just calls import (or dereferences `PyDateTimeAPI`) and then calls the function from PyPy.

Obviously this is under-tested as well, but I think because it involves global state it might be a bigger undertaking to add proper testing to all these functions.

I'm having trouble building the project again (`cannot find -lpypy3-c`), so I'm having trouble testing whether or not this causes problems.

Reply:
> I'm having trouble building the project again (cannot find -lpypy3-c), so I'm having trouble testing whether or not this causes problems.

Have you tried building in a fresh pypy venv? It sometimes seems that a venv works better than `PYTHON_SYS_EXECUTABLE`.

Reply:
@konstin I think it was an issue with conflicts in my `pyenv` setup. I do a lot of library development so my `pyenv` global has a bunch of different things in it. When I set `pyenv shell` to just the pypy version I want it will build, then I just have to adjust `LD_LIBRARY_PATH` to get it to *run*.

In any case, here's an MWE to reproduce the issue, I can confirm it produces a segfault on pypy3.5:

```rust
extern crate pyo3;

use pyo3::prelude::*;
use pyo3::types::PyDateTime;

fn main() {
    let gil = Python::acquire_gil();
    let py = gil.python();
    let x = PyDateTime::from_timestamp(py, 123456.0, None);

}
```

It does not segfault on Python 3.7.2.

Reply:
Ah, my mistake, I can't reproduce the segfault using extension modules, I guess these particular symbols don't rely on the `_PyDateTime_Import` being called.

Reply:
It seems PyPy will import datetime internally when this is called.
```
@cpython_api([PyObject], PyObject)
def PyDateTime_FromTimestamp(space, w_args):
    """Create and return a new datetime.datetime object given an argument tuple
    suitable for passing to datetime.datetime.fromtimestamp().
    """
    w_datetime = PyImport_Import(space, space.newtext("datetime"))
    w_type = space.getattr(w_datetime, space.newtext("datetime"))
    w_method = space.getattr(w_type, space.newtext("fromtimestamp"))
    return space.call(w_method, w_args)
```

Reply:
Yeah, I think `PyDatetime_Import` is needed to initialize the CAPI object, but not when you use the macro directly (but only for PyPy), because of the way it's implemented.

I remember I had some segfaults when I didn't do the import when I was implementing the feature for pypy, which is why I was worried, but I realize now that was because I was dereferencing `PyDatetimeCAPI`, not because I was calling the `fromtimestamp` function.